src/specific/s_filesystem.c

```diff
@@ -35,3 +37,78 @@ void S_File_CreateDirectory(const char *path)
     mkdir(path, 0664);
 #endif
 }
+
+char *S_File_CasePath(char const *path)
+{
+    assert(path);
+
+    char *current_path = Memory_Alloc(strlen(path) + 2);
+    const char *sep = strchr(path, '/') ? "/" : "\\";
+
+    char *path_copy = Memory_DupStr(path);
+    char *path_piece = path_copy;
+
+    bool relative;
+
+    if (path_copy[0] == '/') {
+        relative = false;
+        strcpy(current_path, "/");
+    } else if (strstr(path_copy, ":\\")) {
+        relative = false;
+        strcpy(current_path, path_copy);
+        strstr(current_path, ":\\")[1] = '\0';
+        path_piece += 3;
+    } else {
+        relative = true;
+        strcpy(current_path, ".");
+    }
+
+    while (path_piece) {
+        char *delim = strpbrk(path_piece, "/\\");
+        char old_delim = delim ? *delim : '\0';
+        if (delim) {
+            *delim = '\0';
+        }
+
+        DIR *path_dir = opendir(current_path);
+        if (!path_dir) {
+            Memory_FreePointer(&path_copy);
+            Memory_FreePointer(&current_path);
+            return NULL;
+        }
+
+        struct dirent *cur_file = readdir(path_dir);
+        while (cur_file) {
+            if (strcasecmp(path_piece, cur_file->d_name) == 0) {
+                strcat(current_path, sep);
+                strcat(current_path, cur_file->d_name);
+                break;
+            }
+            cur_file = readdir(path_dir);
+        }
+        closedir(path_dir);
+
+        if (!cur_file) {
+            strcat(current_path, sep);
+            strcat(current_path, path_piece);
+        }
+
+        if (delim) {
+            *delim = old_delim;
+            path_piece = delim + 1;
+        } else {
+            break;
+        }
+    }
+
+finish:
```

Comment:
Is this needed? I don't see any `goto`s.

Reply:
It's not needed